src/inertial_sense_ros.cpp

```diff
@@ -5,857 +5,1227 @@
+    nh_private_.param<bool>("stream_NavSatFix", NavSatFix_.enabled, false);
+    nh_private_.param<bool>("stream_mag", mag_.enabled, false);
+    nh_private_.param<bool>("stream_baro", baro_.enabled, false);
+    nh_private_.param<bool>("stream_preint_IMU", preint_IMU_.enabled, false);
+    nh_private_.param<bool>("stream_diagnostics", diagnostics_.enabled, true);
+
+    SET_CALLBACK(DID_FLASH_CONFIG, nvm_flash_cfg_t, flash_config_callback, 1000);
+
+    if (DID_INS_1_.enabled)
+    {
+        DID_INS_1_.pub = nh_.advertise<inertial_sense_ros::DID_INS1>("DID_INS_1", 1);
+        SET_CALLBACK(DID_INS_1, ins_1_t, INS1_callback, 1);
+    }
+    if (DID_INS_2_.enabled)
+    {
+        DID_INS_2_.pub = nh_.advertise<inertial_sense_ros::DID_INS2>("DID_INS_2", 1);
+        SET_CALLBACK(DID_INS_2, ins_2_t, INS2_callback, 1);
+    }
+    if (DID_INS_4_.enabled)
+    {
+        DID_INS_4_.pub = nh_.advertise<inertial_sense_ros::DID_INS4>("DID_INS_4", 1);
+        SET_CALLBACK(DID_INS_4, ins_4_t, INS4_callback, 1);
+    }
+    if (odom_ins_ned_.enabled)
+    {
+        odom_ins_ned_.pub = nh_.advertise<nav_msgs::Odometry>("odom_ins_ned", 1);
+        SET_CALLBACK(DID_INS_4, ins_4_t, INS4_callback, 1);                                                   // Need NED
+        if (covariance_enabled_)
+            SET_CALLBACK(DID_ROS_COVARIANCE_POSE_TWIST, ros_covariance_pose_twist_t, INS_covariance_callback, 1); // Need Covariance data
+        SET_CALLBACK(DID_PREINTEGRATED_IMU, preintegrated_imu_t, preint_IMU_callback, 1);                     // Need angular rate data from IMU
+        IMU_.enabled = true;
+        // Create Identity Matrix
+        //
+        for (int row = 0; row < 6; row++)
+        {
+            for (int col = 0; col < 6; col++)
+            {
+                if (row == col)
+                {
+                    ned_odom_msg.pose.covariance[row * 6 + col] = 1;
+                    ned_odom_msg.twist.covariance[row * 6 + col] = 1;
+                }
+                else
+                {
+                    ned_odom_msg.pose.covariance[row * 6 + col] = 0;
+                    ned_odom_msg.twist.covariance[row * 6 + col] = 0;
+                }
+            }
+        }
+    }
 
+    if (odom_ins_ecef_.enabled)
+    {
+        odom_ins_ecef_.pub = nh_.advertise<nav_msgs::Odometry>("odom_ins_ecef", 1);
+        SET_CALLBACK(DID_INS_4, ins_4_t, INS4_callback, 1);                                                   // Need quaternion and ecef
+        if (covariance_enabled_)
+            SET_CALLBACK(DID_ROS_COVARIANCE_POSE_TWIST, ros_covariance_pose_twist_t, INS_covariance_callback, 1); // Need Covariance data
+        SET_CALLBACK(DID_PREINTEGRATED_IMU, preintegrated_imu_t, preint_IMU_callback, 1);                                              // Need angular rate data from IMU
+        IMU_.enabled = true;
+        // Create Identity Matrix
+        //
+        for (int row = 0; row < 6; row++)
```

Comment:
SDK ISMatrix.c has a function that creates an identity matrix, **void eye_MatN( f_t *A, i_t n )**. We could probably use it here.

Reply:
OK

Reply:
DIdn't work for doubles